# 1. 데이터 소개

- HIV에 감염된 사람들의 추정 수(모든 연령대)
- HIV/AIDS로 인한 사망자 수
    > Country   
    > Year   
    > Count  
    > Count_median  
    > Count_min  
    > Count_max  
    > WHO Region :국가가 속한 WHO의 지역을 나타냄


    

# 2. 데이터셋 불러오기

In [7]:
# 라이브러리 불러오기
import pandas as pd
import numpy as np

In [268]:
# 데이터셋 불러오기
infected_hiv = pd.read_csv('archive/no_of_people_living_with_hiv_by_country_clean.csv')
dead_hiv = pd.read_csv('archive/no_of_deaths_by_country_clean.csv')

In [17]:
# 데이터셋 확인
display(infected_hiv.head())
print()
display(dead_hiv.head())


Country  Year                  Count  Count_median  Count_min  \
0  Afghanistan  2018       7200[4100–11000]        7200.0     4100.0   
1      Albania  2018                     na           NaN        NaN   
2      Algeria  2018     16000[15000–17000]       16000.0    15000.0   
3       Angola  2018  330000[290000–390000]      330000.0   290000.0   
4    Argentina  2018  140000[130000–150000]      140000.0   130000.0   

   Count_max             WHO Region  
0    11000.0  Eastern Mediterranean  
1        NaN                 Europe  
2    17000.0                 Africa  
3   390000.0                 Africa  
4   150000.0               Americas

Country  Year              Count  Count_median  Count_min  Count_max  \
0  Afghanistan  2018       500[200–610]         500.0      200.0      610.0   
1      Albania  2018                 na           NaN        NaN        NaN   
2      Algeria  2018       200[200–200]         200.0      200.0      200.0   
3       Angola  2018  14000[9500–18000]       14000.0     9500.0    18000.0   
4    Argentina  2018    1700[1300–2100]        1700.0     1300.0     2100.0   

              WHO Region  
0  Eastern Mediterranean  
1                 Europe  
2                 Africa  
3                 Africa  
4               Americas

In [19]:
infected_hiv.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 680 entries, 0 to 679
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Country       680 non-null    object 
 1   Year          680 non-null    int64  
 2   Count         680 non-null    object 
 3   Count_median  553 non-null    float64
 4   Count_min     553 non-null    float64
 5   Count_max     553 non-null    float64
 6   WHO Region    680 non-null    object 
dtypes: float64(3), int64(1), object(3)
memory usage: 37.3+ KB


In [21]:
dead_hiv.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 510 entries, 0 to 509
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Country       510 non-null    object 
 1   Year          510 non-null    int64  
 2   Count         510 non-null    object 
 3   Count_median  400 non-null    float64
 4   Count_min     400 non-null    float64
 5   Count_max     400 non-null    float64
 6   WHO Region    510 non-null    object 
dtypes: float64(3), int64(1), object(3)
memory usage: 28.0+ KB


(메모)  
감염자 수 전체 컬럼수는 680, 사망자 수 전체 컬럼은 510.  데이터를 합친다면 감염자 수 쪽으로 left해야겠음.  
count에 적힌게 추정감[신뢰구간] 으로 적혀있는데, 일단 난 신뢰구간 값을 지워야겠음.

# 3. 데이터 탐색

In [29]:
# count 쪽의 데이터 값들 종류를 보고싶음
infected_hiv['Count'].value_counts() 

Count
na                          127
500[500–500]                 12
100[100–100]                  6
200[200–200]                  6
100[100–200]                  4
                           ... 
480000[400000–560000]         1
1800[1500–2000]               1
17000[13000–21000]            1
1000[930–1100]                1
1600000[1400000–1900000]      1
Name: count, Length: 496, dtype: int64

In [31]:
 # 사망자 쪽의 Count 쪽도 보기
dead_hiv['Count'].value_counts() 

Count
na                      110
100[100–100]            101
500[500–500]             14
500[200–500]             14
100[100–200]             11
                       ... 
22000[17000–27000]        1
10000[7300–13000]         1
1600[1300–2000]           1
2000[980–4100]            1
120000[98000–150000]      1
Name: count, Length: 230, dtype: int64

# 4.  데이터 전처리

- WHO Region 을 mapping 하는게 좋을거 같음
    > replace() 와 map()의 과정이 있음

--- 

'Europe' : 1, 'Africa' : 2 , 'Americas' : 3 ,
'Eastern Mediterranean' :4,
 'Western Pacific' : 5 , 'South-East Asia': 6

In [270]:
infected_hiv['WHO Region'] = infected_hiv['WHO Region'].map({'Europe' : 1, 'Africa' : 2 , 'Americas' : 3 , 'Eastern Mediterranean' :4,
                               'Western Pacific' : 5 , 'South-East Asia': 6})

dead_hiv['WHO Region'] = dead_hiv['WHO Region'].replace({'Europe' : 1, 'Africa' : 2 , 'Americas' : 3 , 'Eastern Mediterranean' :4,
                               'Western Pacific' : 5 , 'South-East Asia': 6})

# 확인
print(dead_hiv['WHO Region'].value_counts())
print()
print(infected_hiv['WHO Region'].value_counts())

WHO Region
1    150
2    132
3     90
4     60
5     48
6     30
Name: count, dtype: int64

WHO Region
1    200
2    176
3    120
4     80
5     64
6     40
Name: count, dtype: int64


C:\Users\User\AppData\Local\Temp\ipykernel_12216\145184587.py:4: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  dead_hiv['WHO Region'] = dead_hiv['WHO Region'].replace({'Europe' : 1, 'Africa' : 2 , 'Americas' : 3 , 'Eastern Mediterranean' :4,


- 결측치 확인 및 처리
    > drop() , dropna() , fillna() , ffill() , bfill(), interpolate()  

In [272]:
# Count에 보면  na라고 적힌 값이 있단 말임. 그 값을 가진 행을 지우기로 함
infected_hiv.drop(infected_hiv[infected_hiv['Count']== 'na'].index, inplace = True)
dead_hiv.drop(dead_hiv[dead_hiv['Count']== 'na'].index, inplace = True)

- 데이터 값 수정

In [274]:
infected_hiv['Count'] = infected_hiv['Count'].apply(lambda x: x.split('[')[0])
dead_hiv['Count'] = dead_hiv['Count'].apply(lambda x: x.split('[')[0])

- 데이터 형태 변경

In [276]:
infected_hiv['Count'] = infected_hiv['Count'].astype(int)
dead_hiv['Count'] = dead_hiv['Count'].astype(int)

- 컬럼 이름 변경
  > rename() , columns 를 이용해서 변경하는 방법이 있음.   
  > rename()의 경우 한,두개 변경시 dict형태로 이름 값을 주어 변경

In [278]:
infected_hiv.columns =['country','year','infected_mean','infected_median','infected_min','infected_max','who_region']
dead_hiv.columns =['country','year','dead_mean','dead_median','dead_min','dead_max','who_region']

In [200]:
infected_hiv.info()

<class 'pandas.core.frame.DataFrame'>
Index: 553 entries, 0 to 679
Data columns (total 7 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   country          553 non-null    object 
 1   year             553 non-null    int64  
 2   infected_mean    553 non-null    int32  
 3   infected_median  553 non-null    float64
 4   infected_min     553 non-null    float64
 5   infected_max     553 non-null    float64
 6   who_region       553 non-null    int64  
dtypes: float64(3), int32(1), int64(2), object(1)
memory usage: 32.4+ KB


- 집계
    > groupby(by=['범주값'],as_index = )['연속값'].연산 메서드() 를 이용하여 집계함
  

In [280]:
hiv_infect_country =infected_hiv.groupby(by=['country'],as_index=False)[['infected_mean',	
                                                                         'infected_median',	'infected_min',	'infected_max']].mean()

hiv_dead_country =dead_hiv.groupby(by=['country'],as_index=False)[['dead_mean',	
                                                                         'dead_median',	'dead_min',	'dead_max']].mean()

- 두 개의 데이터셋을 하나로 합치기
    > concat([df1 , df2 ], axis =  , join = ) : 단순 행,열 기준으로 합치기  
    > merge(df1,df2 , on = , how = ) : 공통된 열을 기준으로 합치기
    > 

In [215]:
print(hiv_infect_country.shape)
print(hiv_dead_country.shape)


(140, 5)
(135, 5)


In [282]:
# country를 기준으로 합병 
hiv = hiv_infect_country.merge(hiv_dead_country , on= 'country' , how = 'left')

# 확인
hiv.head()


country  infected_mean  infected_median  infected_min  infected_max  \
0  Afghanistan         3975.0           3975.0        2325.0        6425.0   
1      Algeria         7175.0           7175.0        6700.0        7650.0   
2       Angola       196750.0         196750.0      165500.0      230000.0   
3    Argentina        99750.0          99750.0       89250.0      108750.0   
4      Armenia         2612.5           2612.5        2095.0        3400.0   

     dead_mean  dead_median     dead_min      dead_max  
0   366.666667   366.666667   166.666667    536.666667  
1   166.666667   166.666667   133.333333    166.666667  
2  9600.000000  9600.000000  6766.666667  12466.666667  
3  1566.666667  1566.666667  1266.666667   1933.333333  
4   166.666667   166.666667   133.333333    166.666667

- 열 순서 변경

In [284]:
hiv = hiv[['country'	,'infected_mean'	,'dead_mean','infected_median','infected_min'
           ,'infected_max','dead_median' ,'dead_min', 'dead_max']]

- 열 추가
    > insert() , 단순 행 추가  
    > insert는 열 인덱스를 기준으로 원하는 위치에 값을 추가할 수 있다.

In [298]:
hiv.drop(columns='mortality_rate',inplace = True)

In [300]:
hiv.insert(loc = 3, column='mortality_rate' , value= ( hiv['dead_mean']/hiv['infected_mean']*100) )


# 확인
hiv.head()

country  infected_mean    dead_mean  mortality_rate  infected_median  \
0  Afghanistan         3975.0   366.666667        9.224319           3975.0   
1      Algeria         7175.0   166.666667        2.322880           7175.0   
2       Angola       196750.0  9600.000000        4.879288         196750.0   
3    Argentina        99750.0  1566.666667        1.570593          99750.0   
4      Armenia         2612.5   166.666667        6.379585           2612.5   

   infected_min  infected_max  dead_median     dead_min      dead_max  
0        2325.0        6425.0   366.666667   166.666667    536.666667  
1        6700.0        7650.0   166.666667   133.333333    166.666667  
2      165500.0      230000.0  9600.000000  6766.666667  12466.666667  
3       89250.0      108750.0  1566.666667  1266.666667   1933.333333  
4        2095.0        3400.0   166.666667   133.333333    166.666667

- 원_핫 인코딩
    > pd.get_dummies()를 이용해서 할 수 있음.  

- 범주화  
    > pd.cut()  -> 내가 구간을 정할 수 있고, 구간 안 원소개수는 다 다름  
    > pd.qcut()  -> 구간 갯수를 입력받으면 각 구간마다 동일한 원소를 가질 수 있음

In [302]:
hiv['level'] = pd.cut(hiv['mortality_rate'],bins = [-np.inf , 3,6 , np.inf] , labels =['low','middle','high'])

#확인
hiv.head()

country  infected_mean    dead_mean  mortality_rate  infected_median  \
0  Afghanistan         3975.0   366.666667        9.224319           3975.0   
1      Algeria         7175.0   166.666667        2.322880           7175.0   
2       Angola       196750.0  9600.000000        4.879288         196750.0   
3    Argentina        99750.0  1566.666667        1.570593          99750.0   
4      Armenia         2612.5   166.666667        6.379585           2612.5   

   infected_min  infected_max  dead_median     dead_min      dead_max   level  
0        2325.0        6425.0   366.666667   166.666667    536.666667    high  
1        6700.0        7650.0   166.666667   133.333333    166.666667     low  
2      165500.0      230000.0  9600.000000  6766.666667  12466.666667  middle  
3       89250.0      108750.0  1566.666667  1266.666667   1933.333333     low  
4        2095.0        3400.0   166.666667   133.333333    166.666667    high

(참고)  
- df['새 변수 이름'] = df['연속형 변수'].rolling(window =(묶음 갯수), min_periods=(최소 몇개부터 계산 적용할 것인가)).연산 메소드.shift()  
- shift()의 경우 입력된 값만큼 행/열을 이동  

- pivot <--> melt